nbgrader/tests/test_nbextension.py

```diff
@@ -0,0 +1,200 @@
+import subprocess as sp
+import tempfile
+import os
+import shutil
+
+from nose.tools import assert_equal
+
+from selenium import webdriver
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support.ui import WebDriverWait
+from selenium.webdriver.support.ui import Select
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.common.keys import Keys
+
+class TestCreateAssignmentNbExtension(object):
+
+    @classmethod
+    def setup_class(cls):
+        cls.tempdir = tempfile.mkdtemp()
+        cls.ipythondir = tempfile.mkdtemp()
+        cls.origdir = os.getcwd()
+        os.chdir(cls.tempdir)
+
+        cls.nbserver = sp.Popen([
+            "ipython", "notebook",
+            "--ipython-dir", cls.ipythondir,
+            "--no-browser",
+            "--port", "9000"], stdout=sp.PIPE, stderr=sp.STDOUT)
+
+    @classmethod
+    def teardown_class(cls):
+        cls.nbserver.kill()
+
+        os.chdir(cls.origdir)
+        shutil.rmtree(cls.tempdir)
+        shutil.rmtree(cls.ipythondir)
+
+    def setup(self):
+        shutil.copy(os.path.join(os.path.dirname(__file__), "files", "blank.ipynb"), "blank.ipynb")
+        self.browser = webdriver.PhantomJS()
```

Comment:
There are two distinct problems we have:

- The old webkit used in Phantom doesn't support everything we want (current websockets protocol, PATCH requests, binary websockets). Being able to use alternative drivers should alleviate this.
- The general async-js awfulness causing things like sporadic failures, unclear error messages when failures do occur, and so on. We're never going to completely escape this, but I'm optimistic that it might be better with Selenium than with Casper, because the difference in language between the tests and the code running requires a cleaner abstraction layer. Also, Selenium seems rather more mature.